ast.ts

```diff
@@ -18,23 +18,39 @@ export type VarInit<A> = { a?: A, name: string, type: Type, value: Literal }
 export type FunDef<A> = { a?: A, name: string, parameters: Array<Parameter<A>>, ret: Type, inits: Array<VarInit<A>>, body: Array<Stmt<A>> }
 
 export type Stmt<A> =
-  | {  a?: A, tag: "assign", name: string, value: Expr<A> }
+  | {  a?: A, tag: "assign", target: Destructure<A>, value: Expr<A> } // TODO: unify field assignment with destructuring. This will eventually replace tag: "id-assign"
+  | {  a?: A, tag: "id-assign", name: string, value: Expr<A> }
   | {  a?: A, tag: "return", value: Expr<A> }
   | {  a?: A, tag: "expr", expr: Expr<A> }
   | {  a?: A, tag: "if", cond: Expr<A>, thn: Array<Stmt<A>>, els: Array<Stmt<A>> }
   | {  a?: A, tag: "while", cond: Expr<A>, body: Array<Stmt<A>> }
   | {  a?: A, tag: "pass" }
   | {  a?: A, tag: "field-assign", obj: Expr<A>, field: string, value: Expr<A> }
 
+export interface Destructure<A> {
+  isDestructured: boolean;
+  targets: AssignTarget<A>[];
+}
+
+export interface AssignTarget<A> {
+  target: Assignable<A>;
+  starred: boolean;
+  ignore: boolean;
+}
+
+// Union of all assignable targets
+export type Assignable<A> =
+  | {  a?: A, tag: "id", name: string }
+  | {  a?: A, tag: "lookup", obj: Expr<A>, field: string }
+
 export type Expr<A> =
     {  a?: A, tag: "literal", value: Literal }
-  | {  a?: A, tag: "id", name: string }
   | {  a?: A, tag: "binop", op: BinOp, left: Expr<A>, right: Expr<A>}
   | {  a?: A, tag: "uniop", op: UniOp, expr: Expr<A> }
   | {  a?: A, tag: "builtin1", name: string, arg: Expr<A> }
   | {  a?: A, tag: "builtin2", name: string, left: Expr<A>, right: Expr<A>}
   | {  a?: A, tag: "call", name: string, arguments: Array<Expr<A>> } 
-  | {  a?: A, tag: "lookup", obj: Expr<A>, field: string }
+  | Assignable<A>
```

Comment:
Okay. seemed similar enough that i originally was valuing trying to keep it DRY. the main reason to keep it as a subset of expr is then we can use the exact same type checking code for these assignment targets (rather than need to write tcAssignTarget(target: Assignable) which duplicates all the type checking logic for lookup, variable, etc.)

Reply:
Hm. I see what you're saying. We could refactor this again later if it's useful. I think that it's useful to keep concepts separate and find some of the overlap later, since we'll have to anyway. For example, I don't want the list group to have to add `l[i] = e` to `Assignable` right now.

Reply:
The main reason for coupling is we need exactly the same parsing and type checking logic for the target of an assignment. In the case of a feature being added—particularly for type checking—it now needs to be rendered in two different places for AST and type checking. This could be seen in inheritance being implemented for example.
for ref: CPython defines `target` type in its own grammar (for our purposes) as a union of `identifier | attributeref | slicing`.

For the time being though, guess i'll just make a copy with a big comment